File: counting_types.py
'''
Count 2413 occurences naively but also collecting all the occurences and their types
Input: permutation as a list
Return: list
        types - list counting type 1, 2, 3 occurences respectively ONLY for horizontal (left-right) type 2 and 3
'''
def count_2413_w_set_type(permutation):
    types = [0, 0, 0]
    P = permutation
    n = len(P)
    for i in range(n-3):
        for j in range(i+1,n-2):
            if P[i] < P[j]:
                for k in range(j+1,n-1):
                    if P[k] < P[i]:
                        for l in range(k+1,n):
                            if P[l] < P[j] and P[l] >  P[i]:
                                occur = [P[i],P[j],P[k],P[l]]
                                if (check_type_1(occur, n)):
                                  types[0] += 1
                                elif (check_type_2(occur, n)):
                                  types[1] += 1
                                elif (check_type_3(occur, n)):
                                  types[2] += 1
    return types

def check_type_1(element, n):
  if (element[0] >= n / 4 and element[1] >= 3 * n / 4 and element[2] < n / 4 and element[3] < 3 * n / 4):
    return True

  return False

def check_type_2(element, n):
  works = 0
  for i in range(len(element)):
    if (element[i] < n / 4 or element[i] >= 3 * n / 4):
      works += 1

  return works == 1

def check_type_3(element, n):
  for i in range(len(element)):
      if (element[i] < n / 4 or element[i] >= 3 * n / 4):
        return False
  return True

File: countingTypes.js
// countingTypes.js
/**
 * Port of count_2413_w_set_type and helpers. :contentReference[oaicite:2]{index=2}
 */
export function count2413WithSetType(P) {
    const types = [0, 0, 0];
    const n = P.length;
  
    for (let i = 0; i < n - 3; i++) {
      for (let j = i + 1; j < n - 2; j++) {
        if (P[i] >= P[j]) continue;
        for (let k = j + 1; k < n - 1; k++) {
          if (P[k] >= P[i]) continue;
          for (let l = k + 1; l < n; l++) {
            if (P[l] < P[j] && P[l] > P[i]) {
              const occ = [P[i], P[j], P[k], P[l]];
              if (checkType1(occ, n)) types[0] += 1;
              else if (checkType2(occ, n)) types[1] += 1;
              else if (checkType3(occ, n)) types[2] += 1;
            }
          }
        }
      }
    }
    return types;
  }
  
  function checkType1(e, n) {
    return e[0] >= n/4 && e[1] >= 3*n/4 && e[2] < n/4 && e[3] < 3*n/4;
  }
  function checkType2(e, n) {
    let edge = 0;
    for (const v of e) if (v < n/4 || v >= 3*n/4) edge += 1;
    return edge === 1;
  }
  function checkType3(e, n) {
    return e.every(v => v >= n/4 && v < 3*n/4);
  }
  